sonar-core/src/main/java/org/sonar/core/issue/tracking/Tracker.java

```diff
@@ -0,0 +1,291 @@
+
+    // 4. match issues with same rule, same line and same message
+    match(tracking, LineAndMessageKeyFactory.INSTANCE);
+
+    // 5. match issues with same rule and same line hash but different line and different message.
+    // See SONAR-2812
+    match(tracking, LineHashKeyFactory.INSTANCE);
+
+    return tracking;
+  }
+
+  private void detectCodeMoves(Input<RAW> rawInput, Input<BASE> baseInput, Tracking<RAW, BASE> tracking) {
+    if (!tracking.isComplete()) {
+      new BlockRecognizer<RAW, BASE>().match(rawInput, baseInput, tracking);
+    }
+  }
+
+  private void match(Tracking<RAW, BASE> tracking, SearchKeyFactory factory) {
+    if (tracking.isComplete()) {
+      return;
+    }
+
+    Multimap<SearchKey, BASE> baseSearch = ArrayListMultimap.create();
+    for (BASE base : tracking.getUnmatchedBases()) {
+      baseSearch.put(factory.create(base), base);
+    }
+
+    Collection<RAW> trackedRaws = new ArrayList<>();
+    for (RAW raw : tracking.getUnmatchedRaws()) {
+      SearchKey rawKey = factory.create(raw);
+      Collection<BASE> bases = baseSearch.get(rawKey);
+      if (!bases.isEmpty()) {
+        // TODO taking the first one. Could be improved if there are more than 2 issues on the same line.
+        // Message could be checked to take the best one.
+        BASE match = bases.iterator().next();
+        tracking.associateRawToBase(raw, match);
+        baseSearch.remove(rawKey, match);
+        trackedRaws.add(raw);
+      }
+    }
+    tracking.markRawsAsAssociated(trackedRaws);
+  }
+
+  private void relocateManualIssues(Input<RAW> rawInput, Input<BASE> baseInput, Tracking<RAW, BASE> tracking) {
+    // FIXME copy of Set if required to avoid concurrent modifications (see tracking.associateManualIssueToLine())
+    Iterable<BASE> manualIssues = from(new HashSet<>(tracking.getUnmatchedBases())).filter(IsManual.INSTANCE);
+    for (BASE base : manualIssues) {
+      if (base.getLine() == null) {
+        // no need to relocate. Location is unchanged.
+        tracking.associateManualIssueToLine(base, null);
+      } else {
+        String baseHash = base.getLineHash();
+        if (Strings.isNullOrEmpty(baseHash)) {
+          baseHash = baseInput.getLineHashSequence().getHashForLine(base.getLine());
+        }
+        if (!Strings.isNullOrEmpty(baseHash)) {
+          int[] rawLines = rawInput.getLineHashSequence().getLinesForHash(baseHash);
+          if (rawLines.length == 1) {
+            tracking.associateManualIssueToLine(base, rawLines[0]);
+          } else if (rawLines.length == 0 && rawInput.getLineHashSequence().hasLine(base.getLine())) {
+            // still valid (???). We didn't manage to correctly detect code move, so the
+            // issue is kept at the same location, even if code changes
+            tracking.associateManualIssueToLine(base, base.getLine());
+          }
+          // TODO if hash found multiple times, , pick the closest line
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)